src/formattedcode/output_cyclonedx.py

```diff
@@ -0,0 +1,659 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) nexB Inc. and others. All rights reserved.
+# ScanCode is a trademark of nexB Inc.
+# SPDX-License-Identifier: Apache-2.0
+# See http://www.apache.org/licenses/LICENSE-2.0 for the license text.
+# See https://github.com/nexB/scancode-toolkit for support or download.
+# See https://aboutcode.org for more information about nexB OSS projects.
+#
+
+import click
+import json
+import re
+import uuid
+from datetime import datetime
+from enum import Enum
+from xml.etree import ElementTree as ET
+
+import attr
+from commoncode.cliutils import OUTPUT_GROUP
+from commoncode.cliutils import PluggableCommandLineOption
+from formattedcode import FileOptionType
+from plugincode.output import OutputPlugin
+from plugincode.output import output_impl
+from typing import FrozenSet, List, Tuple
+
+
+def _get_set_of_known_licenses_and_spdx_license_ids() -> Tuple[List, FrozenSet[str]]:
+    """load all licenses and all SPDX Ids known to ScanCode
+       this will also load scancode licenserefs, so we filter those
+    """
+    from licensedcode.models import get_all_spdx_keys, load_licenses
+    licenses = load_licenses(with_deprecated=True)
+    spdx_keys = filter(lambda x: "LicenseRef" not in x,
+                       get_all_spdx_keys(licenses))
+    return (licenses, frozenset(spdx_keys))
+
+
+known_licenses, spdx_ids = _get_set_of_known_licenses_and_spdx_license_ids()
+
+hash_type_mapping = {
+    'md5': 'MD5',
+    'sha1': 'SHA-1',
+    'sha256': 'SHA-256',
+    'sha512': 'SHA-512',
+}
+
+
+@attr.s
+class CycloneDxLicense:
+    id: str = attr.ib(default=None)
+    name: str = attr.ib(default=None)
+    url: str = attr.ib(default=None)
+
+
+@attr.s
+class CycloneDxLicenseEntry:
+    license: CycloneDxLicense = attr.ib(default=None)
+    expression: str = attr.ib(default=None)
+
+
+@attr.s
+class CycloneDxAttribute:
+    name: str = attr.ib()
+    value: str = attr.ib()
+
+
+@attr.s
+class CycloneDxMetadata():
+    timestamp = attr.ib(
+        default=datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ'))
+    tools: List[dict] = attr.ib(factory=list)
+    properties: List[CycloneDxAttribute] = attr.ib(factory=list)
+
+
+@attr.s
+class CycloneDxHashObject():
+    alg: str = attr.ib()
```

Comment:
may be using a full name doe not hurt, like algorithm? unless this is the cdx name of course.

Reply:
this is the cdx name, so I'd petition to keep it like this so that the default __dict__ repr returns what we want the JSON key to be without modification